plugins/postgres/dbt/adapters/postgres/relation_configs/index.py

```diff
@@ -0,0 +1,174 @@
+from dataclasses import dataclass, field
+from typing import Optional, Set, FrozenSet, Dict, Union, List
+
+from dbt.dataclass_schema import StrEnum
+from dbt.exceptions import DbtRuntimeError
+from dbt.adapters.relation_configs import (
+    RelationConfigBase,
+    RelationResults,
+    RelationConfigValidationMixin,
+    RelationConfigValidationRule,
+    RelationConfigChangeAction,
+    RelationConfigChange,
+)
+
+
+# A `NodeConfig` instance can have multiple indexes, this is just one index
+# e.g. {"columns": ["column_a", "column_b"], "unique": True, "type": "hash"}
+Columns = List[str]
+ModelNodeEntry = Dict[str, Union[Columns, bool, str]]
+
+
+class PostgresIndexMethod(StrEnum):
+    btree = "btree"
+    hash = "hash"
+    gist = "gist"
+    spgist = "spgist"
+    gin = "gin"
+    brin = "brin"
+
+    @classmethod
+    def default(cls) -> "PostgresIndexMethod":
+        return cls.btree
+
+
+@dataclass(frozen=True, eq=True, unsafe_hash=True)
+class PostgresIndexConfig(RelationConfigBase, RelationConfigValidationMixin):
+    """
+    This config fallows the specs found here:
+    https://www.postgresql.org/docs/current/sql-createindex.html
+
+    The following parameters are configurable by dbt:
+    - name: the name of the index in the database, this isn't predictable since we apply a timestamp
+    - unique: checks for duplicate values when the index is created and on data updates
+    - method: the index method to be used
+    - column_names: the columns in the index
+
+    Applicable defaults for non-configurable parameters:
+    - concurrently: `False`
+    - nulls_distinct: `True`
+    """
+
+    name: Optional[str] = field(default=None, hash=False, compare=False)
+    column_names: Optional[FrozenSet[str]] = field(default_factory=set, hash=True)
+    unique: Optional[bool] = field(default=False, hash=True)
+    method: Optional[PostgresIndexMethod] = field(default=PostgresIndexMethod.btree, hash=True)
+
+    @property
+    def validation_rules(self) -> Set[RelationConfigValidationRule]:
+        return {
+            RelationConfigValidationRule(
+                validation_check=self.column_names is not None,
+                validation_error=DbtRuntimeError(
+                    "Indexes require at least one column, but none were provided"
+                ),
+            ),
+        }
+
+    @classmethod
+    def from_dict(cls, config_dict) -> "PostgresIndexConfig":
+        kwargs_dict = {
+            "name": config_dict.get("name"),
+            "method": config_dict.get("method"),
+            "unique": config_dict.get("unique"),
+            "column_names": frozenset(column for column in config_dict.get("column_names", {})),
+        }
+        index: "PostgresIndexConfig" = super().from_dict(kwargs_dict)  # type: ignore
+        return index
+
+    @classmethod
+    def parse_model_node(cls, model_node_entry: ModelNodeEntry) -> dict:
```

Comment:
looks like this method and `parse_relation_results` are both returning intermediate dictionaries that eventually get passed to `from_dict`. What do you think about refactoring them both to be `from_model_node` and `from_relation_results` respectively - each calling `from_dict` under the hood. That way the caller has less to worry about when initializing a `PostgresIndexConfig`.

Reply:
hmm, looks like something similar has been done for `PostgresMaterializedViewConfig`.

Reply:
I did that so that `MaterializedView.parse_model_node()` returns a standard dictionary, and not a dictionary with `RelationConfig` objects. More specifically, I did that so that `MaterializedView.from_dict()` can accept a standard dictionary, in case there are other things that want to use this that don't want to build the `IndexConfig` objects manually before submitting them to `MaterializedView.from_dict()`. Ultimately this might all be replaced with something like `mashumaro` or `pydantic` under the hood, so I want `from_dict()` to look as close as possible to methods in those frameworks. I could build the `from_model_node()` and `from_relation_results()` methods for `IndexConig`; I just didn't need them, so I didn't build them.